contracts/MultiPartyEscrow.sol

```diff
@@ -93,23 +95,53 @@ contract MultiPartyEscrow {
         require(balances[msg.sender] >= value, "Insufficient balance in the contract.");
         require(signer != address(0));
 
+        require(_openChannel(msg.sender, signer, recipient, groupId, value, expiration), "Unable to open channel");
+        return true;
+    }
+    
+    //open a channel on behalf of the user. Sender should send the signed permission to open the channel
+    function openChannelByThirdParty(address sender, address signer, address recipient, bytes32 groupId, uint256 value, uint256 expiration, uint256 messageNonce, uint8 v, bytes32 r, bytes32 s) 
+    public
+    returns(bool) 
+    {
+        require(balances[sender] >= value, "Insufficient balance");
+
+        require(messageNonce >= block.number-5 && messageNonce <= block.number+5, "Invalid message nonce");
+
+        //compose the message which was signed
+        bytes32 message = prefixed(keccak256(abi.encodePacked(this, msg.sender, signer, recipient, groupId, value, expiration, messageNonce)));
+        
+        //check for replay attack (message can be used only once)
+        require( ! usedMessages[message], "Invalid signature");
```

Comment:
I propose to change error message to something like this  "This message has already been used"

Reply:
Will change the error message to "Signature has already been used".
In general I prefer not to reveal underlying validations to the end users. That's why gave it as a general error message.